### PR TITLE
pQueue

### PR DESCRIPTION
**Pull request by jnware7**

### CODE REVIEW

File: src/pQueue.js [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+  }
+
+  isEmpty() {
+    return this.q.length <= 0
+  }
+
+  lengths() {
+    return this.q.length 
+  }
+}
```
Comment (devonwesley) [inline src/pQueue.js:37]: Everything looks good to me. But like I said to y'all in person. I think the built .length method is cheating. You should try and implement your own. But if you understand then go ahead and merge.

**Comment by devonwesley** [comment]
After you look and my comment go ahead and merge.

File: spec/pQueue.js [hunk 2]
```diff
@@ -0,0 +1,91 @@
+
+import chai, { expect } from 'chai'
+import chaiChange from 'chai-change'
+import PriorityQueue from '../src/pQueue'
+
+chai.use(chaiChange)
+
+// beforeEach(function() {
```
Comment (jaredatron) [inline spec/pQueue.js:8]: try not to commit these comments

File: spec/pQueue.js [hunk 3]
```diff
@@ -0,0 +1,91 @@
+
```
Comment (jaredatron) [inline spec/pQueue.js:1]: remove this blank line at the top of the file

File: spec/pQueue.js [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+  })
+
+  context('enqueue()', () => {
+    it('adds an element with priority (number) to the back of the queue', () => {
+      const pQueue = new PriorityQueue()
+      pQueue.enqueue('dude', 90)
+      pQueue.enqueue('cool', 20)
+      pQueue.enqueue('turd', 50)
+      pQueue.enqueue('sal', 36)
+      expect(pQueue.enqueue('foo',70)).to.eql([{element: 'dude', priority: 90}, {element: 'foo', priority: 70},{element: 'turd', priority: 50}, {element: 'cool', priority: 20}])
```
Comment (jaredatron) [inline spec/pQueue.js:26]: this line is too long:

```js
expect(pQueue.enqueue('foo',70)).to.eql([
  {element: 'dude', priority: 90},
  {element: 'foo',  priority: 70},
  {element: 'turd', priority: 50},
  {element: 'cool', priority: 20},
])
```
is much more readable

File: src/pQueue.js [hunk 5]
```diff
@@ -0,0 +1,37 @@
+'use strict'
+export default class pQueue {
+  constructor() {
+    this.q = []
+  }
+
+  enqueue(element, priority){
+    // push an object to the end of the arrat
+    // sort the array according to priority
+    this.q[this.q.length] = {element, priority}
```
Comment (jaredatron) [inline src/pQueue.js:10]: this is the same as:
```js
this.q.push({element, priority})
```

File: src/pQueue.js [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+
+    var sorted = this.q.sort(function(a,b){
+      return parseFloat(b.priority) - parseFloat(a.priority)
```
Comment (jaredatron) [inline src/pQueue.js:13]: I believe `parseFloat` requires a second argument specifying base10 math `return parseFloat(b.priority, 10) - parseFloat(a.priority, 10)`

File: src/pQueue.js [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+    })
+    return sorted
```
Comment (jaredatron) [inline src/pQueue.js:15]: just return instead of `var sorted = ...; return sorted`

File: src/pQueue.js [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+  }
+
+  front() {
+    return this.q.length > 0 ? this.q[0] : null
```
Comment (jaredatron) [inline src/pQueue.js:19]:
```js
return this.q[0] || null
```

File: src/pQueue.js [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+  }
+
+  back() {
+    return this.q.length > 0 ? this.q[this.q.length - 1] : null
```
Comment (jaredatron) [inline src/pQueue.js:23]:
```js
return this.q[this.q.length - 1] || null
```

File: src/pQueue.js (hunk 6 above)
Comment (jaredatron) [reply src/pQueue.js:13]: why not `parseFloat` the priority when you add it to the array so you dont have to do it a ton of tines during the sort?

File: src/pQueue.js [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+  }
+
+  dequeue() {
+    return this.q.length > 0 ? this.q.shift() : null
+  }
+
+  isEmpty() {
+    return this.q.length <= 0
```
Comment (jaredatron) [inline src/pQueue.js:31]: length should never be less than 0

File: src/pQueue.js [hunk 11]
```diff
@@ -0,0 +1,37 @@
...
+  front() {
+    return this.q.length > 0 ? this.q[0] : null
+  }
+
+  back() {
+    return this.q.length > 0 ? this.q[this.q.length - 1] : null
+  }
+
+  dequeue() {
+    return this.q.length > 0 ? this.q.shift() : null
```
Comment (jaredatron) [inline src/pQueue.js:27]:
```js
return this.isEmpty() ? this.q.shift() : null
```

**Review by jaredatron — COMMENTED** [review]
CC: @Jusdev89

**Pull request merged**